dvc/repo/tree.py

```diff
@@ -127,15 +134,86 @@ def isexec(self, path):
 class RepoTree(BaseTree):
     def __init__(self, repo):
         self.repo = repo
-        self.dvctree = DvcTree(repo)
+        if isinstance(repo, Repo):
+            self.dvctree = DvcTree(repo)
+        else:
+            # git-only erepo's do not need dvctree
+            self.dvctree = None
 
-    def open(self, *args, **kwargs):
-        try:
-            return self.dvctree.open(*args, **kwargs)
-        except FileNotFoundError:
-            pass
+    def open(self, path, **kwargs):
+        if self.dvctree and self.dvctree.exists(path):
+            try:
+                return self.dvctree.open(path, **kwargs)
+            except FileNotFoundError:
+                pass
 
-        return self.repo.tree.open(*args, **kwargs)
+        return self.repo.tree.open(path, **kwargs)
 
     def exists(self, path):
-        return self.repo.tree.exists(path) or self.dvctree.exists(path)
+        return self.repo.tree.exists(path) or (
+            self.dvctree and self.dvctree.exists(path)
+        )
+
+    def isdir(self, path):
+        return self.repo.tree.isdir(path) or (
+            self.dvctree and self.dvctree.isdir(path)
+        )
+
+    def isdvc(self, path):
+        return self.dvctree and self.dvctree.isdvc(path)
+
+    def isfile(self, path):
+        return self.repo.tree.isfile(path) or (
+            self.dvctree and self.dvctree.isfile(path)
+        )
+
+    def _walk(self, top, topdown=True):
+        if self.dvctree and not self.repo.tree.isdir(top):
+            yield from self.dvctree.walk(top, topdown=topdown)
+            return
+        if not self.dvctree or not self.dvctree.isdir(top):
+            yield from self.repo.tree.walk(top, topdown=topdown)
+            return
+
+        # walk and merge both trees, ensure that dvcfiles are ignored (handled
+        # as DVC outs, not as git versioned files)
+        repo_root, repo_dirs, repo_files = first(
+            self.repo.tree.walk(top, topdown=topdown)
+        )
+        dvc_root, dvc_dirs, dvc_files = first(
+            self.dvctree.walk(top, topdown=topdown)
+        )
```

Comment:
Is this efficient enough?

Reply:
This is more of a placeholder while I'm getting the rest of the erepo behavior to work, but eventually, `RepoTree.walk` should only `repo.tree.walk` and `dvctree.walk` once, and do the proper merging/recursion with the initial set of iterators returned from each walk function